Cache NERVES_DL_DIR to S3 to improve build performance

mobileoverlord opened the pull request:
The download directory caching has broken for since moving to Circle CI. This is due to how circle handles caches. Essentially, their caching mechanism will only write a cache file if one does not already exist. This does not work for how we want to cache and share the download directory. This PR updates the Circle CI config to save and restore the contents of the download directory to a file on S3 in the bucket `nerves-project/nerves_system_rpi3/ci-dl-cache.tar.gz`. Operations reading from or writing to the cache will not fail the job if the file is unavailable or there was an issue. Instead it will just try.

mobileoverlord commented:
We have tried something similar to this in the past which caused races between parallel builds trying to read from the cache file while another container was writing to it. S3 seems to have read after write consistency for some time now so I believe this _shouldn't_ be an issue.

fhunleth commented:
I don't think that I understand why CircleCI's caching mechanism doesn't work. If Buildroot downloads `/nerves/dl/newfile.tar.gz`, then `newfile.tar.gz` won't get copied to the cache or will not get restored?

mobileoverlord commented:
Here is a snippet from CircleCI's docs regarding caching
> Cache is immutable on write: once a cache is written for a particular key like node-cache-master, it cannot be written to again. Consider a workflow of 3 jobs, where Job3 depends on Job1 and Job2: {Job1, Job2} -> Job3. They all read and write to the same cache key.
>
> In a run of the workflow, Job3 may use the cache written by Job1 or Job2. Since caches are immutable, this would be whichever job saved its cache first.

Under these limitations I don't see a way to offload the cache cost to circle that is meaningful.

fhunleth commented:
Ahh, so caches are treated as one big blob by CircleCI. That's a bummer. 

How about we use Buildroot's caching mechanism to have a place to get the tar files individually? We could s3 sync the dl directory with the cache. I'd expect that most of the time nothing new would be downloaded so the sync would be pretty quick. Also, CI would just download the files it needs. I can show you more on how this works on slack.

mobileoverlord commented:
I have updated this PR with a couple improvements

* Added BR2_PRIMARY_SITE and pointed it to s3://nerves-project/dl
* aws sync the contents of the dl directory back to s3://nerves-project/dl after a successful build.
* Remove build directory caching. Currently, the build directory is being added to a circle ci cache to be shared with the `build_test` job. It takes quite a while to package up the build directory. In stead of sharing the build directory, we are unpacking the already cached artifacts into the `build_test` job containers `NERVES_DL_DIR`.
* Wait to install the aws cli client until just before its needed. This is a fast fail optimization.

Overall this shaves ~ 10 min off the build time and ~1 min off the build_test time

fhunleth commented on .circleci/config.yml:
```diff
@@ -28,6 +28,26 @@ install_nerves_bootstrap: &install_nerves_bootstrap
+sync_dl_cache: &sync_dl_cache
+  run: 
+    name: Sync downloads cache
+    command: aws s3 sync ~/.nerves/dl s3://nerves-project/dl || true
```
`--storage-class REDUCED_REDUNDANCY`?

fhunleth commented on nerves_defconfig:
```diff
@@ -1,6 +1,7 @@
 BR2_arm=y
 BR2_cortex_a53=y
 BR2_ARM_FPU_NEON_VFPV4=y
+BR2_PRIMARY_SITE="https://s3.amazonaws.com/nerves-project/dl"
```
I have mixed feelings about everyone hitting our S3 bucket first when they build. Let's chat about this. Maybe it's ok.
